File: Counting-Sort/python/leetcode_lc_1122.py
arr1 = [2,3,1,3,2,4,6,7,9,2,19]
arr2 = [2,1,4,3,9,6]

class Solution:
    def relativeSortArray(self, arr1, arr2):
        
        for i in range(len(arr1)):
            if i == 0:
                max = arr1[i]
            if arr1[i] > max:
                max = arr1[i]

        bucket = self.countingHandler(arr1,max)
        idx = 0
        for i in range(len(arr2)):
            while bucket[arr2[i]]:
                arr1[idx] = arr2[i]
                idx+=1
                bucket[arr2[i]] -= 1

        for j in range(len(bucket)):
            while bucket[j]:
                arr1[idx] = j
                idx += 1
                bucket[j] -= 1
        
        return arr1


    def countingHandler(self,arr,max):
        bucket = [0]*(max+1)
        for i in range(len(arr)):
            if not bucket[arr[i]]:
                bucket[arr[i]] = 1
            else:
                bucket[arr[i]] += 1


        return bucket


solution = Solution()

print("arr1:{} \narr2:{} \n{} ".format(arr1,arr2,solution.relativeSortArray(arr1,arr2)))

File: Counting-Sort/js/leetcode_lc_1122.js
const arr1 = [2,21,43,38,0,42,33,7,24,13,12,27,12,24,5,23,29,48,30,31]
const arr2 = [2,42,38,0,43,21]

/**
 * @param {number[]} arr1
 * @param {number[]} arr2
 * @return {number[]}
 */
var relativeSortArray = function(arr1, arr2) {
    let max
    for(let i = 0 ; i < arr1.length ; i++){
        if(i===0){
            max = arr1[0]
        }
        if(arr1[i] > max){
            max = arr1[i]
        }
    }
    const bucket = countingSort(arr1,max)

    let idx = 0
    for(let i = 0 ; i < arr2.length ; i++){
        while(bucket[arr2[i]]){
            arr1[idx] = arr2[i]
            idx++
            bucket[arr2[i]] -= 1
        }
    }
//  console.log(arr1)
    bucket.forEach((v,index) => {
        // console.log("==>",v,index)
        while(v){
            arr1[idx] = index
            idx++
            v--
        }
    })

    // console.log(arr1)
    return arr1
};

function countingSort(arr,max){
    const bucket = new Array(max+1)
    let idx = 0
    for(let i = 0 ; i< arr.length;i++){
        if(String(bucket[arr[i]]) === "undefined"){
            bucket[arr[i]] = 1
        }else{
            bucket[arr[i]] += 1
        }
    }

    return bucket
}


console.log(`arr1:[${arr1}] \narr2:[${arr2}] \n ${relativeSortArray(arr1,arr2)}`)